File: services/api/app/models/regression.py
"""Regression test models"""
from datetime import datetime
from typing import Optional, List, Dict, Any
from sqlalchemy import Column, Integer, String, Text, Float, Boolean, DateTime, ForeignKey, JSON
from sqlalchemy.orm import relationship
from pydantic import BaseModel

from ..db import Base


# ==================== SQLAlchemy Models ====================

class RegressionTestCase(Base):
    """Test case definition for regression testing"""
    __tablename__ = 'regression_test_cases'
    
    id = Column(Integer, primary_key=True)
    name = Column(String(200), nullable=False)
    category = Column(String(50), nullable=False)  # rag/llm/e2e
    query = Column(Text, nullable=False)
    expected_ku_ids = Column(JSON, default=list)
    expected_answer = Column(Text)
    evaluation_criteria = Column(JSON, default=dict)
    tags = Column(JSON, default=list)
    is_active = Column(Boolean, default=True)
    created_by = Column(String(50))
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    results = relationship("RegressionTestResult", back_populates="case")


class RegressionTestRun(Base):
    """Test execution run"""
    __tablename__ = 'regression_test_runs'
    
    id = Column(Integer, primary_key=True)
    run_id = Column(String(50), unique=True, nullable=False)
    status = Column(String(20), default='running')  # running/completed/failed
    total_cases = Column(Integer, default=0)
    passed_cases = Column(Integer, default=0)
    failed_cases = Column(Integer, default=0)
    review_cases = Column(Integer, default=0)
    pass_rate = Column(Float)
    started_at = Column(DateTime, default=datetime.utcnow)
    completed_at = Column(DateTime)
    triggered_by = Column(String(50))
    config = Column(JSON, default=dict)
    
    # Relationships
    results = relationship("RegressionTestResult", back_populates="run", cascade="all, delete-orphan")


class RegressionTestResult(Base):
    """Individual test result"""
    __tablename__ = 'regression_test_results'
    
    id = Column(Integer, primary_key=True)
    run_id = Column(Integer, ForeignKey('regression_test_runs.id', ondelete='CASCADE'), nullable=False)
    case_id = Column(Integer, ForeignKey('regression_test_cases.id', ondelete='CASCADE'), nullable=False)
    actual_answer = Column(Text)
    retrieved_ku_ids = Column(JSON, default=list)
    retrieval_score = Column(Float)
    answer_score = Column(Float)
    llm_evaluation = Column(JSON)
    manual_review = Column(String(20), default='pending')  # pending/pass/fail
    manual_comment = Column(Text)
    reviewed_by = Column(String(50))
    reviewed_at = Column(DateTime)
    execution_time_ms = Column(Integer)
    status = Column(String(20), default='pending')  # pass/fail/review/pending
    error_message = Column(Text)
    created_at = Column(DateTime, default=datetime.utcnow)
    
    # Relationships
    run = relationship("RegressionTestRun", back_populates="results")
    case = relationship("RegressionTestCase", back_populates="results")


# ==================== Pydantic Schemas ====================

class TestCaseCreate(BaseModel):
    name: str
    category: str  # rag/llm/e2e
    query: str
    expected_ku_ids: List[str] = []
    expected_answer: Optional[str] = None
    evaluation_criteria: Dict[str, Any] = {}
    tags: List[str] = []
    is_active: bool = True


class TestCaseUpdate(BaseModel):
    name: Optional[str] = None
    category: Optional[str] = None
    query: Optional[str] = None
    expected_ku_ids: Optional[List[str]] = None
    expected_answer: Optional[str] = None
    evaluation_criteria: Optional[Dict[str, Any]] = None
    tags: Optional[List[str]] = None
    is_active: Optional[bool] = None


class TestCaseResponse(BaseModel):
    id: int
    name: str
    category: str
    query: str
    expected_ku_ids: List[str]
    expected_answer: Optional[str]
    evaluation_criteria: Dict[str, Any]
    tags: List[str]
    is_active: bool
    created_by: Optional[str]
    created_at: Optional[str]
    updated_at: Optional[str]
    
    class Config:
        from_attributes = True


class TestRunCreate(BaseModel):
    category_filter: Optional[str] = None  # Filter by category
    tag_filter: Optional[List[str]] = None  # Filter by tags
    case_ids: Optional[List[int]] = None  # Specific case IDs to run


class TestRunResponse(BaseModel):
    id: int
    run_id: str
    status: str
    total_cases: int
    passed_cases: int
    failed_cases: int
    review_cases: int
    pass_rate: Optional[float]
    started_at: Optional[str]
    completed_at: Optional[str]
    triggered_by: Optional[str]
    
    class Config:
        from_attributes = True


class LLMEvaluation(BaseModel):
    accuracy: float
    completeness: float
    relevance: float
    clarity: float
    overall: float
    comment: str


class TestResultResponse(BaseModel):
    id: int
    run_id: int
    case_id: int
    case_name: Optional[str] = None
    case_query: Optional[str] = None
    case_category: Optional[str] = None
    actual_answer: Optional[str]
    retrieved_ku_ids: List[str]
    retrieval_score: Optional[float]
    answer_score: Optional[float]
    llm_evaluation: Optional[Dict[str, Any]]
    manual_review: str
    manual_comment: Optional[str]
    reviewed_by: Optional[str]
    reviewed_at: Optional[str]
    execution_time_ms: Optional[int]
    status: str
    error_message: Optional[str]
    created_at: Optional[str]
    
    class Config:
        from_attributes = True


class ManualReviewRequest(BaseModel):
    review: str  # pass/fail
    comment: Optional[str] = None


class TestCaseImport(BaseModel):
    cases: List[TestCaseCreate]

File: services/admin-ui/src/api/regression.ts
import { apiClient } from './client'

// ==================== Types ====================

export interface TestCase {
  id: number
  name: string
  category: 'rag' | 'llm' | 'e2e'
  query: string
  expected_ku_ids: string[]
  expected_answer?: string
  evaluation_criteria: Record<string, unknown>
  tags: string[]
  is_active: boolean
  created_by?: string
  created_at?: string
  updated_at?: string
}

export interface TestCaseCreate {
  name: string
  category: 'rag' | 'llm' | 'e2e'
  query: string
  expected_ku_ids?: string[]
  expected_answer?: string
  evaluation_criteria?: Record<string, unknown>
  tags?: string[]
  is_active?: boolean
}

export interface TestCaseUpdate {
  name?: string
  category?: 'rag' | 'llm' | 'e2e'
  query?: string
  expected_ku_ids?: string[]
  expected_answer?: string
  evaluation_criteria?: Record<string, unknown>
  tags?: string[]
  is_active?: boolean
}

export interface TestRun {
  id: number
  run_id: string
  status: 'running' | 'completed' | 'failed'
  total_cases: number
  passed_cases: number
  failed_cases: number
  review_cases: number
  pass_rate?: number
  started_at?: string
  completed_at?: string
  triggered_by?: string
}

export interface TestRunCreate {
  category_filter?: string
  tag_filter?: string[]
  case_ids?: number[]
}

export interface LLMEvaluation {
  accuracy?: number
  completeness?: number
  relevance?: number
  clarity?: number
  overall?: number
  comment?: string
  faithfulness?: number
  answer_relevancy?: number
  context_precision?: number
  context_recall?: number
  [key: string]: unknown
}

export interface TestResult {
  id: number
  run_id: number
  case_id: number
  case_name?: string
  case_query?: string
  case_category?: string
  actual_answer?: string
  retrieved_ku_ids: string[]
  retrieval_score?: number
  answer_score?: number
  llm_evaluation?: LLMEvaluation
  manual_review: 'pending' | 'pass' | 'fail'
  manual_comment?: string
  reviewed_by?: string
  reviewed_at?: string
  execution_time_ms?: number
  status: 'pass' | 'fail' | 'review' | 'pending'
  error_message?: string
  created_at?: string
}

export interface ManualReview {
  review: 'pass' | 'fail'
  comment?: string
}

export interface RegressionStats {
  total_cases: number
  cases_by_category: Record<string, number>
  total_runs: number
  avg_pass_rate: number
  pending_reviews: number
  recent_runs: TestRun[]
}

export interface EvaluatorHealth {
  evaluator_healthy: boolean
  evaluator_url: string
}

// ==================== API Functions ====================

export const regressionApi = {
  // Test Cases
  listCases: async (params?: {
    category?: string
    tag?: string
    is_active?: boolean
    skip?: number
    limit?: number
  }): Promise<TestCase[]> => {
    const response = await apiClient.get('/api/v1/regression/cases', { params })
    return response.data
  },

  getCase: async (caseId: number): Promise<TestCase> => {
    const response = await apiClient.get(`/api/v1/regression/cases/${caseId}`)
    return response.data
  },

  createCase: async (data: TestCaseCreate): Promise<TestCase> => {
    const response = await apiClient.post('/api/v1/regression/cases', data)
    return response.data
  },

  updateCase: async (caseId: number, data: TestCaseUpdate): Promise<TestCase> => {
    const response = await apiClient.put(`/api/v1/regression/cases/${caseId}`, data)
    return response.data
  },

  deleteCase: async (caseId: number): Promise<void> => {
    await apiClient.delete(`/api/v1/regression/cases/${caseId}`)
  },

  importCases: async (cases: TestCaseCreate[]): Promise<{
    imported: number
    total: number
    errors: Array<{ index: number; error: string }>
  }> => {
    const response = await apiClient.post('/api/v1/regression/cases/import', { cases })
    return response.data
  },

  // Test Runs
  listRuns: async (params?: {
    status?: string
    skip?: number
    limit?: number
  }): Promise<TestRun[]> => {
    const response = await apiClient.get('/api/v1/regression/runs', { params })
    return response.data
  },

  getRun: async (runId: number): Promise<TestRun> => {
    const response = await apiClient.get(`/api/v1/regression/runs/${runId}`)
    return response.data
  },

  createRun: async (data: TestRunCreate): Promise<TestRun> => {
    const response = await apiClient.post('/api/v1/regression/runs', data)
    return response.data
  },

  getRunResults: async (runId: number, params?: {
    status?: string
    manual_review?: string
  }): Promise<TestResult[]> => {
    const response = await apiClient.get(`/api/v1/regression/runs/${runId}/results`, { params })
    return response.data
  },

  // Manual Review
  submitReview: async (resultId: number, data: ManualReview): Promise<{
    message: string
    status: string
  }> => {
    const response = await apiClient.post(`/api/v1/regression/results/${resultId}/review`, data)
    return response.data
  },

  // Statistics
  getStats: async (): Promise<RegressionStats> => {
    const response = await apiClient.get('/api/v1/regression/stats')
    return response.data
  },

  // Evaluator Health
  checkEvaluatorHealth: async (): Promise<EvaluatorHealth> => {
    const response = await apiClient.get('/api/v1/regression/evaluator/health')
    return response.data
  },
}

export default regressionApi

